File: csphn_ore.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

const int MAXN = 500;
int gold[MAXN][MAXN], silver[MAXN][MAXN];
int prefix_gold[MAXN][MAXN], prefix_silver[MAXN][MAXN];
int max_gold_row[MAXN][MAXN], max_silver_col[MAXN][MAXN];

int main() {
    int m, n;
    cin >> m >> n;

    // Input for gold and silver
    for (int i = 0; i < m; i++)
        for (int j = 0; j < n; j++)
            cin >> gold[i][j];

    for (int i = 0; i < m; i++)
        for (int j = 0; j < n; j++)
            cin >> silver[i][j];

    // Compute prefix sums for gold (leftward sums)
    for (int i = 0; i < m; i++) {
        prefix_gold[i][0] = gold[i][0];
        for (int j = 1; j < n; j++)
            prefix_gold[i][j] = prefix_gold[i][j-1] + gold[i][j];
    }

    // Compute prefix sums for silver (upward sums)
    for (int j = 0; j < n; j++) {
        prefix_silver[0][j] = silver[0][j];
        for (int i = 1; i < m; i++)
            prefix_silver[i][j] = prefix_silver[i-1][j] + silver[i][j];
    }

    // Calculate max gold row-wise
    for (int i = 0; i < m; i++) {
        max_gold_row[i][0] = prefix_gold[i][0];
        for (int j = 1; j < n; j++) {
            max_gold_row[i][j] = max(max_gold_row[i][j-1], prefix_gold[i][j]);
        }
    }

    // Calculate max silver column-wise
    for (int j = 0; j < n; j++) {
        max_silver_col[0][j] = prefix_silver[0][j];
        for (int i = 1; i < m; i++) {
            max_silver_col[i][j] = max(max_silver_col[i-1][j], prefix_silver[i][j]);
        }
    }

    // Calculate maximum value
    int max_value = 0;
    for (int i = 0; i < m; i++) {
        for (int j = 0; j < n; j++) {
            // Max value from gold ending at (i,j) and silver not overlapping
            int value = max_gold_row[i][j] + max_silver_col[i][j];
            max_value = max(max_value, value);
        }
    }

    // Output the result
    cout << max_value << endl;

    return 0;
}

File: csphn_ore.py
def max_ore_value(m, n, gold_values, silver_values):
    gold = [[0] * n for _ in range(m)]
    silver = [[0] * n for _ in range(m)]

    # Initialize the values for the last column and last row
    for i in range(m):
        gold[i][n-1] = gold_values[i][n-1]
    for j in range(n):
        silver[m-1][j] = silver_values[m-1][j]

    # Update the values for the remaining cells
    for i in range(m):
        for j in range(n-2, -1, -1):
            gold[i][j] = gold_values[i][j] + gold[i][j+1]
    for j in range(n):
        for i in range(m-2, -1, -1):
            silver[i][j] = silver_values[i][j] + silver[i+1][j]

    # Calculate the optimal value
    max_value = 0
    for i in range(m):
        for j in range(n):
            max_value = max(max_value, gold[i][j] + silver[i][j])

    return max_value

import sys
input = sys.stdin.read
data = input().split()

# Read the input values
m = int(data[0])
n = int(data[1])
gold_values = []
silver_values = []

index = 2
for i in range(m):
    gold_values.append(list(map(int, data[index:index + n])))
    index += n

for i in range(m):
    silver_values.append(list(map(int, data[index:index + n])))
    index += n

# Calculate the result
result = max_ore_value(m, n, gold_values, silver_values)

# Print the result
print(result)
